proposals/NNNN-ComparisonReform.md

```diff
@@ -0,0 +1,531 @@
+traditional definition of “comparison” for these types. Take, for example,
+sorting an array of `Float`s. Today, `Float`'s instance of `Comparable` follows
+IEEE-754 and returns `false` for all comparisons of `NaN`. In order to sort
+this array, `NaN` s are considered outside the domain of `<`, and the order of
+a sorted array containing them is unspecified. Similarly, a Dictionary keyed
+off floats can leak entries and memory.
+
+2: Generic algorithms in the Swift Standard Library that make use of the
+current `Comparable` protocol may have to make extra comparisons to determine
+the ordering of values when `<`, `==`, and `>` should have different
+behaviours. Having a central operation to return complete ordering information
+should provide a speedup for these operations.
+
+3: The existing comparison operators don't "generalize" well. There's no clean
+way to add a third or fourth argument to `<` to ask for non-default semantics.
+An example where this would be desirable would be specifying the locale or
+case-sensitivity when comparing Strings.
+
+4: `Comparable` is over-engineered in the customization points it provides: to
+our knowledge, there's no good reason to ever override `>=`, `>`, or `<=`. Each
+customization point bloats vtables and mandates additional dynamic dispatch.
+
+5: When quickly writing a `Comparable` type, it is easier to implement a single
+ternary statement than to separately implement `==` and `<`.
+
+## Proposed solution
+
+### `ComparisonResult`
+
+Foundation's ComparisonResult type will be mapped into Swift as 
+
+```swift
+@objc public enum ComparisonResult : Int {
+  case orderedAscending = -1
+  case orderedSame = 0
+  case orderedDescending = 1
+}
+```
+
+
+### `Comparable`
+
+Comparable will be changed to have a new ternary comparison method: `compare(_
+other: Self) -> ComparisonResult`. `x.compare(y)` specifies where to place x
+relative to y. So if it yields `.orderedAscending`, then x comes before y. This
+will be considered the new "main" dispatch point of Comparable that
+implementors should provide.
+
+Most code will continue to use `<` or `==`, as it will be optimal for their
+purposes. However code that needs to make a three-way branch on comparison can
+use the potentially more efficient `compare`. Note that `compare` is only
+expected to be more efficient in this specific case. If a two-way branch is all
+that's being done, `<` will be more efficient in many cases (if only because
+it's easier for the optimizer).
+
+For backwards compatibility reasons, `compare` will have a default
+implementation defined in terms of `<`, but to enable only using `compare`,
+`<` and `==` will also have default implementations in terms of `compare`.
+
+The compiler will verify that either `compare`, or `<` and `==`, are provided
+by every type that claims to conform to `Comparable`. This will be done in some
+unspecified way unavailable outside the standard library (it can be made
```

Comment:
I'm slightly uncomfortable giving magic behavior to standard library protocols, but I agree that this is the behavior we want…